### DISCUSSION TITLE
Are multiple training sessions cumulative?

### DISCUSSION THREAD

**Discussion by richard-a-lott** [moj-analytical-services/splink#1480] [Q&A] [answered]
A pretty simple question, I hope, which I'm not entirely clear on; Are training sessions of the same type cumulative?

For u parameter training, if I run 2 or more random sampling sessions in a row, will the final u parameters be a combined result of both sessions, or will they be overwritten each time?

Same for m parameter training; I need to run several EM training sessions to ensure all parameters are filled, but can I run _extra_ sessions to get better predictions, or will the existing m parameters be overwritten by new ones?

**Comment by RobinL** [comment, marked as answer]
Yes.  Once way to see this is with:
```
linker.estimate_u_using_random_sampling(target_rows=1e4)
linker.estimate_u_using_random_sampling(target_rows=1e4)
linker.estimate_u_using_random_sampling(target_rows=1e4)

linker.parameter_estimate_comparisons_chart()
```

<img width="630" alt="image" src="https://github.com/moj-analytical-services/splink/assets/2608005/1f2ae547-0824-4047-8b4b-e375a7e504d3">


You'll see that the parameter estimate chart contains the different estimates and if you then run 
```
linker.m_u_parameters_chart()
```

you'll see the value chosen is the average of the values in `parameter_estimate_comparisons_chart()`

<details><summary>fuller working script</summary>

```python
# from try_settings import settings_dict
import pandas as pd

from splink.duckdb.comparison_library import (
    exact_match,
    levenshtein_at_thresholds,
)
from splink.duckdb.linker import DuckDBLinker

df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")



settings = {
    "probability_two_random_records_match": 0.01,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.first_name = r.first_name",
        "l.surname = r.surname",
    ],
    "comparisons": [
        levenshtein_at_thresholds("first_name", 2),
        exact_match("surname"),
        exact_match("dob"),
        exact_match("city", term_frequency_adjustments=True),
        exact_match("email"),
    ],
    "retain_matching_columns": False,
    "retain_intermediate_calculation_columns": False,
    "additional_columns_to_retain": ["group"],
    "max_iterations": 10,
    "em_convergence": 0.01,
}




linker = DuckDBLinker(df, settings, connection=":memory:")


linker.estimate_u_using_random_sampling(target_rows=1e6)
linker.estimate_u_using_random_sampling(target_rows=1e6)
linker.estimate_u_using_random_sampling(target_rows=1e6)



blocking_rule = "l.first_name = r.first_name and l.surname = r.surname"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule)


blocking_rule = "l.dob = r.dob"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule)

linker.parameter_estimate_comparisons_chart()
linker.m_u_parameters_chart()

```

</details>
<img width="1045" alt="image" src="https://github.com/moj-analytical-services/splink/assets/2608005/4422a0a2-fdda-494e-a384-eaaa2181ae27">

**Reply by RobinL** [reply]
Also worth noting that EM can be a bit fickle for estimating m, so if you get any values which are counterintuitive in the `parameter_estimate_comparisons_chart()` you could just drop that training session or replace it with a different one that gives more sensible values.

**Reply by richard-a-lott** [reply]
Thanks that's really helpful! I wasn't aware I could produce that top chart! I should take a closer look at what else the Linker can do!

One related question. I have a feature in my data which is very poorly populated but is an important ID that I need in the model. I'm finding that random u estimation is problematic because there are rarely random comparisons falling into my exact or 'near' match, and Splink tells me my model is not fully trained. I'm currently using a loop which trains u and checks whether u is fully populated, before continuing. Is this the best way, or is there another alternative?

**Reply by RobinL** [reply]
The best option if it's possible is to set `target_rows` larger, e.g. we typically use 1bn, `1e9`

If you only get reasonable estimates occasionally, you could just hard code these known values following running estimate_u.  It's a bit fiddly to do, but the syntax is like:
```
linker._settings_obj.comparisons[1].comparison_levels[1].u_probability = 1e-8
```
obviously you need your indexes on comparisons and comparisons levels to match the comparison level you're setting

You could also set the values manually if you know (at least roughly) what the values _should_ be  i.e. don't bother estimating, just set by expert judgement

**Reply by richard-a-lott** [reply]
I should have said that my loop is doing 1bn a time. It's taking forever! I like the hard coding approach though. I'll give that a try. Thanks again.

**Reply by RobinL** [reply]
Another trick you could use is setting up a model with a single comparison on your important id, and run estimate u on an even larger target rows.  It'll run faster since I'll it only needs to do estimates for that one comparison.  Once you have the values, you can hard code them during other runs as above

**Reply by richard-a-lott** [reply]
Ah nice idea! I'll also give that a try! Cheers
